server/storage/rethinkdb.go

```diff
@@ -0,0 +1,205 @@
+func (r RDBTUFFile) TableName() string {
+	return "tuf_files"
+}
+
+// RDBKey is the public key record
+type RDBKey struct {
+	rethinkdb.Timing
+	Gun    string `gorethink:"gun"`
+	Role   string `gorethink:"role"`
+	Cipher string `gorethink:"cipher"`
+	Public []byte `gorethink:"public"`
+}
+
+// TableName returns the table name for the record type
+func (r RDBKey) TableName() string {
+	return "tuf_keys"
+}
+
+// RethinkDB implements a MetaStore against the Rethink Database
+type RethinkDB struct {
+	dbName string
+	sess   *gorethink.Session
+}
+
+// NewRethinkDBStorage initializes a RethinkDB object
+func NewRethinkDBStorage(dbName string, sess *gorethink.Session) RethinkDB {
+	return RethinkDB{
+		dbName: dbName,
+		sess:   sess,
+	}
+}
+
+// GetKey returns the cipher and public key for the given GUN and role.
+// If the GUN+role don't exist, returns an error.
+func (rdb RethinkDB) GetKey(gun, role string) (cipher string, public []byte, err error) {
+	var key RDBKey
+	res, err := gorethink.DB(rdb.dbName).Table(key.TableName()).GetAllByIndex(
+		rdbGunRoleIdx, []string{gun, role},
+	).Run(rdb.sess)
+	if err != nil {
+		return "", nil, err
+	}
+	defer res.Close()
+	err = res.One(&key)
+	if err == gorethink.ErrEmptyResult {
+		return "", nil, &ErrNoKey{gun: gun}
+	}
+	return key.Cipher, key.Public, err
+}
+
+// SetKey sets the cipher and public key for the given GUN and role if
+// it doesn't already exist.  Otherwise an error is returned.
+func (rdb RethinkDB) SetKey(gun, role, cipher string, public []byte) error {
+	now := time.Now()
+	key := RDBKey{
+		Timing: rethinkdb.Timing{
+			CreatedAt: now,
+			UpdatedAt: now,
+		},
+		Gun:    gun,
+		Role:   role,
+		Cipher: cipher,
+		Public: public,
+	}
+	_, err := gorethink.DB(rdb.dbName).Table(key.TableName()).Insert(key).RunWrite(rdb.sess)
+	return err
+}
+
+// UpdateCurrent adds new metadata version for the given GUN if and only
+// if it's a new role, or the version is greater than the current version
+// for the role. Otherwise an error is returned.
+func (rdb RethinkDB) UpdateCurrent(gun string, update MetaUpdate) error {
+	now := time.Now()
+	checksum := sha256.Sum256(update.Data)
+	file := RDBTUFFile{
+		Timing: rethinkdb.Timing{
+			CreatedAt: now,
+			UpdatedAt: now,
+		},
+		GunRoleVersion: []interface{}{gun, update.Role, update.Version},
+		Gun:            gun,
+		Role:           update.Role,
+		Version:        update.Version,
+		Sha256:         hex.EncodeToString(checksum[:]),
+		Data:           update.Data,
+	}
+	_, err := gorethink.DB(rdb.dbName).Table(file.TableName()).Insert(
+		file,
+		gorethink.InsertOpts{
+			Conflict: "error", // default but explicit for clarity of intent
+		},
+	).RunWrite(rdb.sess)
+	if err != nil && gorethink.IsConflictErr(err) {
+		return &ErrOldVersion{}
+	}
+	return err
+}
+
+// UpdateMany adds multiple new metadata for the given GUN. RethinkDB does
+// not support transactions, therefore we will attempt to insert the timestamp
```

Comment:
IIRC were we going to insert the timestamp last so that the timestamp ends up being the second phase of the two-phase commit?  Otherwise get the latest entries via consistent walking, we might end up getting a version that hasn't fully been committed/might have conflicts?

I may also just be missing it, as I haven't looked at the rest of the rethink code yet, but this doesn't seem to be pulling out the timestamp update to be inserted at the beginning or end.